src/ServiceControl.Monitoring/QueueLength/QueueLengthCalculator.cs

```diff
@@ -0,0 +1,101 @@
+﻿namespace ServiceControl.Monitoring.QueueLength
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Http;
+    using Newtonsoft.Json.Linq;
+    using Raw;
+
+    public class QueueLengthCalculator : IRawDataConsumer, IEndpointDataProvider
+    {
+        static readonly JArray EmptyArray = new JArray();
+        const string Name = "QueueLength";
+
+        public QueueLengthCalculator()
+        {
+            sessionSends = new ConcurrentDictionary<Guid, long>();
+            sessionReceives = new ConcurrentDictionary<string, ConcurrentDictionary<Guid, long>>();
+            Current = sessionReceives.Select(kvp =>
+            {
+                long totalCount = 0;
+                foreach (var sessionCounterReceive in kvp.Value)
+                {
+                    var sessionId = sessionCounterReceive.Key;
+                    var count = sessionCounterReceive.Value;
+
+                    long sendValue;
+                    if (sessionSends.TryGetValue(sessionId, out sendValue))
+                    {
+                        var diff = sendValue - count;
+                        if (diff > 0)
+                        {
+                            // do not report negative (sender has a lagging reporter)
+                            totalCount += diff;
+                        }
+                    }
+                }
+
+                return new KeyValuePair<string, JObject>(kvp.Key, new JObject(new JProperty("Count", totalCount)));
+            });
+        }
+
+        public void Consume(IReadOnlyDictionary<string, string> headers, JObject data)
+        {
+            var endpoint = headers.GetOriginatingEndpoint();
+
+            var counters = (JArray)data["Counters"] ?? EmptyArray;
+            var sends = counters.Where(c => c.Value<string>("Name").StartsWith("QueueLengthSend_"));
+            UpdateSends(sends);
+
+            var gauges = (JArray)data["Gauges"] ?? EmptyArray;
+            var receives = gauges.Where(c => c.Value<string>("Name").StartsWith("QueueLengthReceive_"));
+            UpdateReceives(receives, endpoint);
+        }
+
+        void UpdateSends(IEnumerable<JToken> sends)
+        {
+            foreach (var send in sends)
+            {
+                var name = send.Value<string>("Name");
+                var sessionId = Guid.Parse(name.Substring(name.LastIndexOf("_") + 1));
```

Comment:
I'd say, let's stick to the semantics of tags: https://github.com/etishor/Metrics.NET/wiki/Grouping-&-Organizing-Metrics#tags and put the "QueueLengthSend" or "Receive" in there, keeping the session id in the name of the counter/gauge.